src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -540,5 +555,22 @@ private static bool IsRequiredImplicitNumericConversion(ITypeSymbol sourceType,
                     return false;
             }
         }
+
+        private static bool IsInUncheckedContext(CastExpressionSyntax cast)
+        {
+            SyntaxNode currentNode = cast;
+
+            do
+            {
+                if (currentNode.Kind() == SyntaxKind.UncheckedExpression || currentNode.Kind() == SyntaxKind.UncheckedStatement)
+                {
+                    return true;
+                }
+
+                currentNode = currentNode.Parent;
+            } while (currentNode is ExpressionSyntax || currentNode is StatementSyntax);
```

Comment:
@tmat Regarding this: _The context can also be set by a compilation wide option._

The default compilation option is unchecked for non-constant expressions. I don't think want to disable all explicit numeric cast simplifications at project level. Probably this is better addressed post 1.0, I'll add this to the work item I create.